src/Silicon.cpp

```diff
@@ -825,70 +935,273 @@ namespace galsim {
         // Now we add in the tree ring distortions
         addTreeRingDistortions(target, orig_center);
 
-        // Start with the correct distortions for the initial image as it is already
-        dbg<<"Initial updatePixelDistortions\n";
-        updatePixelDistortions(target);
-
         // Keep track of the charge we are accumulating on a separate image for efficiency
         // of the distortion updates.
         _delta.resize(b);
         _delta.setZero();
+
+
+        // convert and map data for GPU
+        double* deltaData = _delta.getData();
+        int imageDataSize = _delta.getMaxPtr() - _delta.getData();
+
+        T* targetData = target.getData();
+        
+        int pixelInnerBoundsSize = _pixelInnerBounds.size();
+
+        int hbpSize = _horizontalBoundaryPoints.size();
+        int vbpSize = _verticalBoundaryPoints.size();
+
+        int hdSize = _horizontalDistortions.size();
+        int vdSize = _verticalDistortions.size();
+        
+        // this will only be fully accurate for cases where the table uses linear
+        // interpolation, and the data points are evenly spaced. Currently this is
+        // always the case for _abs_length_table.
+        _abs_length_arg_min = _abs_length_table.argMin();
+        _abs_length_arg_max = _abs_length_table.argMax();
+        _abs_length_size = _abs_length_table.size();
+
+        _abs_length_table_GPU.resize(_abs_length_size);
+        _abs_length_increment = (_abs_length_arg_max - _abs_length_arg_min) /
+            (double)(_abs_length_size - 1);
+        for (int i = 0; i < _abs_length_size; i++) {
+            _abs_length_table_GPU[i] =
+                _abs_length_table.lookup(_abs_length_arg_min + (((double)i) * _abs_length_increment));
+        }
+
+        double* abs_length_table_data = _abs_length_table_GPU.data();
+        
+        int emptypolySize = _emptypoly.size();
+        _emptypolyGPU.resize(emptypolySize);
+        for (int i = 0; i < emptypolySize; i++) {
+            _emptypolyGPU[i].x = _emptypoly[i].x;
+            _emptypolyGPU[i].y = _emptypoly[i].y;
+        }
+        Position<double>* emptypolyData = _emptypolyGPU.data();
+        
+        int nxny = nx * ny;
+        _changed = std::shared_ptr<bool>(new bool[nxny]);
```

Comment:
This only plausibly works with C++17, not C++11, which is our nominal minimum requirement.  And even in C++17, this isn't quite right.  It should be `std::shared_ptr<bool[]>`.  Otherwise it won't delete properly.  cf. https://kezunlin.me/post/b82753fc/

My recommendation is to use `default_delete<bool[]>` as the simplest portable version.  Or possibly switch to `unique_ptr`, since supposedly it has overloads for array deletion.  But you still need to use `bool[]` for the template type.

Reply:
Good point, I missed that subtlety. I suspect it worked fine in practice because `bool` is a primitive type and doesn't need anything special for array deletion, but I've replaced it with `std::unique_ptr<bool[]>` to be safe.